### DISCUSSION TITLE
Combining useWaitForTransaction & useTransaction hooks

### DISCUSSION THREAD

**Discussion by milaabl** [wevm/wagmi#2789] [Ideas]
Is combining useWaitForTransaction & useTransaction hooks possible? With the current behavior, I have to do the following to show the transaction details both while it's loading and when it's completed:

```typescript
const {
    error,
    data: transactionData,
    isLoading: isTransactionDataLoading,
  } = useWaitForTransaction({
    hash: transactionHash,
    enabled: !!(transactionHash),
    onError: (err) => console.log({ err }),
  });

  const {
    data: transactionDetails
  } = useTransaction({
    hash: transactionHash,
    enabled: !!(transactionHash && !isTransactionDataLoading),
    onError: (err) => console.log({ err }),
  });
```

Please advice!

**Comment by Seranged** [comment]
+1 this thread

I understand you guys want to make it more decomposable with each hook, but it is annoying to have to manually generate the hooks for `useContractWrite` and `useWaitForTransaction` to create a full flow for the UX on a button that is alternating its render based on the TX state. Forgive me if there is a easier way of creating this and I haven't found it

e.g.

tx click handler
tx submit
tx pending
tx confirmed

Is there any chance of adding in some of the props from useWaitForTransaction into useContractWrite so we can remove the necessity of `useWaitForTransaction (wagmi)` / `waitForTransactionReceipt (viem)`? It would make the code much cleaner for end-user developers

Thank you for your hard work guys 😊

**Reply by jxom** [reply]
Can you not compose wagmi Hooks into your own [Custom Hook](https://react.dev/learn/reusing-logic-with-custom-hooks)?

Here is my analogy applied to wagmi from the React docs:

> React comes with several built-in Hooks like useState, useContext, and useEffect. Sometimes, you’ll wish that there was a Hook for some more specific purpose: for example, to fetch data, to keep track of whether the user is online, or to connect to a chat room. You might not find these Hooks in React, but you can create your own Hooks for your application’s needs.

wagmi comes with several built-in Hooks like useContractWrite and useWaitForTransaction. Sometimes, you’ll wish there was a Hook for a more specific purpose: for example, an end-to-end transaction flow. You might not find these Hooks in wagmi, but you can create your own Hooks for your application’s needs.

**Reply by Seranged** [reply]
Yes I tend to agree. What i'll do is generate a custom hook then just pull it into repositories that I need it in. 

Thank you for your response!
